File: android/app/src/main/java/com/pehchaanrnscaffold/FaceRecognitionModule.kt
package com.pehchaanrnscaffold

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.RectF
import android.util.Base64
import com.facebook.react.bridge.*
import org.json.JSONArray
import org.tensorflow.lite.Interpreter
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.*

/**
 * React Native bridge for offline face recognition and liveness detection.
 *
 * Models bundled in assets/:
 *   mobilefacenet_indian.tflite — INT8, 512-d embeddings, Indian demographic fine-tuned
 *   blazeface.tflite            — float32, 128×128 input, 896 anchors × 16 regressors
 *
 * JS interface (NativeModules.FaceRecognition):
 *   runInference(faceFrameBase64, candidatesJson, threshold) → NativeInferenceOutput
 *   checkLiveness(framesBase64[], challenge) → LivenessResult
 */
class FaceRecognitionModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        const val MODULE_NAME = "FaceRecognition"
        private const val MFN_SIZE = 112       // MobileFaceNet input size
        private const val BF_SIZE = 128        // BlazeFace input size
        private const val EMBED_DIM = 512      // MobileFaceNet embedding dimension
        private const val BF_SCORE_THRESH = 0.5f
        private const val BF_SCALE = 128f      // BlazeFace decode scale = input size
        private const val BLINK_EAR_THRESH = 0.25f
        private const val BLINK_MIN_FRAMES = 2
        private const val YAW_TURN_THRESH = 15f
        private const val YAW_FRAME_RATIO = 0.4f
    }

    override fun getName() = MODULE_NAME

    // Lazy-loaded interpreters — allocated only when first called
    private val mfnInterpreter: Interpreter by lazy { loadModel("mobilefacenet_indian.tflite") }
    private val bfInterpreter: Interpreter by lazy { loadModel("blazeface.tflite") }
    private val bfAnchors: List<FloatArray> by lazy { generateBlazeFaceAnchors() }
    private var bfRegressorsOutputIdx = 0
    private var bfScoresOutputIdx = 1

    private fun loadModel(fileName: String): Interpreter {
        val fd = reactApplicationContext.assets.openFd(fileName)
        val bytes = fd.createInputStream().use { it.readBytes() }
        fd.close()
        val buf = ByteBuffer.allocateDirect(bytes.size).apply {
            order(ByteOrder.nativeOrder())
            put(bytes)
            rewind()
        }
        val interp = Interpreter(buf, Interpreter.Options().apply { setNumThreads(4) })
        // For BlazeFace: identify output indices by tensor shape
        if (fileName.contains("blazeface")) {
            for (i in 0 until interp.outputTensorCount) {
                val shape = interp.getOutputTensor(i).shape()
                if (shape.size == 3 && shape[2] == 16) bfRegressorsOutputIdx = i
                if (shape.size == 3 && shape[2] == 1) bfScoresOutputIdx = i
            }
        }
        return interp
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Face Recognition
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Run face recognition against a list of enrolled workers.
     *
     * @param faceFrameBase64  Base64 JPEG/PNG of the face crop (any size, will be resized to 112×112)
     * @param candidatesJson   JSON array of WorkerEmbeddingEntry objects
     * @param threshold        Minimum cosine similarity to report a match (e.g. 0.30)
     * @param promise          Resolves with NativeInferenceOutput
     */
    @ReactMethod
    fun runInference(
        faceFrameBase64: String,
        candidatesJson: String,
        threshold: Double,
        promise: Promise,
    ) {
        try {
            val t0 = System.currentTimeMillis()

            // Decode image
            val bytes = Base64.decode(faceFrameBase64, Base64.DEFAULT)
            val src = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                ?: throw IllegalArgumentException("Could not decode image")
            val bmp = Bitmap.createScaledBitmap(src, MFN_SIZE, MFN_SIZE, true)

            // Quantisation params from model (avoids hardcoding scale/zp)
            val inputTensor = mfnInterpreter.getInputTensor(0)
            val scale = inputTensor.quantizationParams().scale.takeIf { it > 0f } ?: 0.00784f
            val zp = inputTensor.quantizationParams().zeroPoint

            // Build INT8 input buffer: NHWC, normalised to [-1, 1] then quantised
            val inputBuf = ByteBuffer.allocateDirect(MFN_SIZE * MFN_SIZE * 3)
                .apply { order(ByteOrder.nativeOrder()) }
            for (y in 0 until MFN_SIZE) {
                for (x in 0 until MFN_SIZE) {
                    val px = bmp.getPixel(x, y)
                    inputBuf.put(quantize(Color.red(px) / 127.5f - 1f, scale, zp))
                    inputBuf.put(quantize(Color.green(px) / 127.5f - 1f, scale, zp))
                    inputBuf.put(quantize(Color.blue(px) / 127.5f - 1f, scale, zp))
                }
            }
            inputBuf.rewind()

            // Run MobileFaceNet
            val output = Array(1) { FloatArray(EMBED_DIM) }
            mfnInterpreter.run(inputBuf, output)
            val queryEmb = output[0]
            val queryNorm = l2norm(queryEmb)

            // Compare against each candidate
            var bestId: String? = null
            var bestScore = 0f
            val candidates = JSONArray(candidatesJson)
            for (i in 0 until candidates.length()) {
                val entry = candidates.getJSONObject(i)
                if (entry.optBoolean("isRevoked", false)) continue
                val embBytes = Base64.decode(entry.getString("embeddingBase64"), Base64.DEFAULT)
                val embFloat = FloatArray(embBytes.size / 4)
                ByteBuffer.wrap(embBytes).order(ByteOrder.LITTLE_ENDIAN)
                    .asFloatBuffer().get(embFloat)
                val score = cosine(queryEmb, queryNorm, embFloat)
                if (score > bestScore) {
                    bestScore = score
                    bestId = entry.getString("workerId")
                }
            }

            val inferenceMs = System.currentTimeMillis() - t0
            promise.resolve(Arguments.createMap().apply {
                putString("workerId", if (bestScore >= threshold.toFloat()) bestId else null)
                putDouble("confidence", bestScore.toDouble())
                putDouble("inferenceMs", inferenceMs.toDouble())
                putDouble("qualityScore", 1.0)
            })
        } catch (e: Exception) {
            promise.reject("INFERENCE_ERROR", e.message ?: "Unknown error", e)
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Liveness Detection
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Evaluate a liveness challenge across a sequence of camera frames.
     *
     * @param framesBase64  ReadableArray of base64 JPEG frames (full camera resolution)
     * @param challenge     "blink" | "turn_left" | "turn_right"
     * @param promise       Resolves with LivenessResult (passed, ear?, yawDegrees?, durationMs)
     */
    @ReactMethod
    fun checkLiveness(
        framesBase64: ReadableArray,
        challenge: String,
        promise: Promise,
    ) {
        try {
            val t0 = System.currentTimeMillis()
            val earSeries = mutableListOf<Float>()
            val yawSeries = mutableListOf<Float>()

            for (i in 0 until framesBase64.size()) {
                val imgBytes = Base64.decode(framesBase64.getString(i)!!, Base64.DEFAULT)
                val bmp = BitmapFactory.decodeByteArray(imgBytes, 0, imgBytes.size) ?: continue
                val kps = runBlazeFace(bmp) ?: continue
                earSeries.add(kps.earProxy)
                yawSeries.add(kps.yawDegrees)
            }

            val durationMs = System.currentTimeMillis() - t0
            val result = Arguments.createMap()
            result.putDouble("durationMs", durationMs.toDouble())

            when (challenge) {
                "blink" -> {
                    val passed = detectBlink(earSeries)
                    result.putBoolean("passed", passed)
                    result.putDouble("ear", earSeries.averageOrZero())
                }
                "turn_left" -> {
                    val passed = yawSeries.count { it < -YAW_TURN_THRESH } >= (yawSeries.size * YAW_FRAME_RATIO)
                    result.putBoolean("passed", passed)
                    result.putDouble("yawDegrees", yawSeries.averageOrZero())
                }
                "turn_right" -> {
                    val passed = yawSeries.count { it > YAW_TURN_THRESH } >= (yawSeries.size * YAW_FRAME_RATIO)
                    result.putBoolean("passed", passed)
                    result.putDouble("yawDegrees", yawSeries.averageOrZero())
                }
                else -> result.putBoolean("passed", false)
            }

            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("LIVENESS_ERROR", e.message ?: "Unknown error", e)
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // BlazeFace inference + keypoint decoding
    // ─────────────────────────────────────────────────────────────────────────

    private data class FaceKeypoints(val earProxy: Float, val yawDegrees: Float)

    private fun runBlazeFace(src: Bitmap): FaceKeypoints? {
        val bmp = Bitmap.createScaledBitmap(src, BF_SIZE, BF_SIZE, true)

        // Float32 input [1, 128, 128, 3] in [0, 1]
        val inputBuf = ByteBuffer.allocateDirect(BF_SIZE * BF_SIZE * 3 * 4)
            .apply { order(ByteOrder.nativeOrder()) }
        for (y in 0 until BF_SIZE) {
            for (x in 0 until BF_SIZE) {
                val px = bmp.getPixel(x, y)
                inputBuf.putFloat(Color.red(px) / 255f)
                inputBuf.putFloat(Color.green(px) / 255f)
                inputBuf.putFloat(Color.blue(px) / 255f)
            }
        }
        inputBuf.rewind()

        // Outputs: regressors [1,896,16], scores [1,896,1]
        val regressors = Array(1) { Array(896) { FloatArray(16) } }
        val scores = Array(1) { Array(896) { FloatArray(1) } }
        val outputs = mapOf<Int, Any>(
            bfRegressorsOutputIdx to regressors,
            bfScoresOutputIdx to scores,
        )
        bfInterpreter.runForMultipleInputsOutputs(arrayOf<Any>(inputBuf), outputs)

        // Pick best anchor above threshold
        var bestScore = BF_SCORE_THRESH
        var bestIdx = -1
        for (i in 0 until 896) {
            val s = sigmoid(scores[0][i][0])
            if (s > bestScore) { bestScore = s; bestIdx = i }
        }
        if (bestIdx < 0) return null

        return decodeKeypoints(regressors[0][bestIdx], bfAnchors[bestIdx])
    }

    /**
     * Decode raw BlazeFace regressor into face keypoints.
     *
     * Keypoint order: [right_eye, left_eye, nose_tip, mouth, right_ear, left_ear]
     * Encoding: alternating [y_offset, x_offset] pairs after the 4 bbox values.
     *
     * EAR proxy: ratio of (eye_y to mouth_y) distance vs face height.
     *   Open eyes → ~0.35–0.50. Closed eyes → ~0.20–0.28.
     *
     * Yaw estimate: nose_x offset from eye midpoint, normalised by face width.
     *   Straight ahead → ~0°. Turn left → negative. Turn right → positive.
     */
    private fun decodeKeypoints(raw: FloatArray, anchor: FloatArray): FaceKeypoints {
        val anchorCx = anchor[0]
        val anchorCy = anchor[1]
        val faceW = (raw[3] / BF_SCALE).coerceAtLeast(0.01f)
        val faceH = (raw[2] / BF_SCALE).coerceAtLeast(0.01f)

        val kpX = FloatArray(6)
        val kpY = FloatArray(6)
        for (i in 0 until 6) {
            kpX[i] = raw[4 + i * 2 + 1] / BF_SCALE + anchorCx
            kpY[i] = raw[4 + i * 2]     / BF_SCALE + anchorCy
        }

        // EAR proxy: vertical eye-to-mouth distance relative to face height
        val eyeMidY = (kpY[0] + kpY[1]) / 2f
        val mouthY = kpY[3]
        val earProxy = ((mouthY - eyeMidY) / faceH).coerceIn(0f, 1f)

        // Yaw: nose offset from eye midpoint, mapped to degrees
        val eyeMidX = (kpX[0] + kpX[1]) / 2f
        val noseX = kpX[2]
        val yawDegrees = ((noseX - eyeMidX) / faceW) * 90f

        return FaceKeypoints(earProxy = earProxy, yawDegrees = yawDegrees)
    }

    private fun detectBlink(earSeries: List<Float>): Boolean {
        if (earSeries.size < BLINK_MIN_FRAMES) return false
        var runLength = 0
        var maxRun = 0
        for (v in earSeries) {
            if (v < BLINK_EAR_THRESH) { runLength++; maxRun = maxOf(maxRun, runLength) }
            else runLength = 0
        }
        return maxRun >= BLINK_MIN_FRAMES
    }

    // ─────────────────────────────────────────────────────────────────────────
    // BlazeFace anchor generation
    // Short-range 128×128: strides [8, 16], anchors-per-cell [2, 6] → 896 total
    // ─────────────────────────────────────────────────────────────────────────

    private fun generateBlazeFaceAnchors(): List<FloatArray> {
        val strides = intArrayOf(8, 16)
        val apc = intArrayOf(2, 6)
        val anchors = ArrayList<FloatArray>(896)
        for (s in strides.indices) {
            val cells = BF_SIZE / strides[s]
            for (row in 0 until cells) {
                for (col in 0 until cells) {
                    val cx = (col + 0.5f) / cells
                    val cy = (row + 0.5f) / cells
                    repeat(apc[s]) { anchors.add(floatArrayOf(cx, cy)) }
                }
            }
        }
        return anchors  // size = 16*16*2 + 8*8*6 = 512 + 384 = 896
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Math helpers
    // ─────────────────────────────────────────────────────────────────────────

    private fun quantize(v: Float, scale: Float, zp: Int): Byte =
        (v / scale + zp).roundToInt().coerceIn(-128, 127).toByte()

    private fun l2norm(v: FloatArray): Float {
        var s = 0f; for (x in v) s += x * x; return sqrt(s).coerceAtLeast(1e-10f)
    }

    private fun cosine(a: FloatArray, aNorm: Float, b: FloatArray): Float {
        val bNorm = l2norm(b)
        var dot = 0f
        for (i in 0 until minOf(a.size, b.size)) dot += a[i] * b[i]
        return (dot / (aNorm * bNorm)).coerceIn(-1f, 1f)
    }

    private fun sigmoid(x: Float) = 1f / (1f + exp(-x))

    private fun List<Float>.averageOrZero() = if (isEmpty()) 0.0 else average()
}

File: ios/PehchaanRNScaffold/FaceRecognitionModule.swift
import Foundation
import UIKit
import TensorFlowLite

/**
 * React Native bridge for offline face recognition and liveness detection (iOS).
 * Mirrors FaceRecognitionModule.kt exactly — same method signatures, same logic.
 *
 * Models loaded from app bundle:
 *   mobilefacenet_indian.tflite — INT8, 512-d embeddings
 *   blazeface.tflite            — float32, 128×128 input
 */
@objc(FaceRecognitionModule)
class FaceRecognitionModule: NSObject {

    // MARK: - Constants

    private let mfnSize = 112
    private let bfSize  = 128
    private let embedDim = 512
    private let bfScoreThresh: Float = 0.5
    private let blinkEarThresh: Float = 0.25
    private let yawTurnThresh: Float = 15.0
    private let yawFrameRatio: Float = 0.4

    // MARK: - Lazy interpreters

    private lazy var mfnInterpreter: Interpreter? = loadModel("mobilefacenet_indian")
    private lazy var bfInterpreter: Interpreter?  = loadModel("blazeface")
    private lazy var bfAnchors: [[Float]] = generateBlazeFaceAnchors()
    private var bfRegressorsIdx = 0
    private var bfScoresIdx = 1

    private func loadModel(_ name: String) -> Interpreter? {
        guard let path = Bundle.main.path(forResource: name, ofType: "tflite") else {
            return nil
        }
        var options = Interpreter.Options()
        options.threadCount = 4
        let interp = try? Interpreter(modelPath: path, options: options)
        try? interp?.allocateTensors()

        // Identify BlazeFace output order by shape
        if name.contains("blazeface"), let interp = interp {
            for i in 0 ..< (try? interp.outputTensorCount()).map({ Int($0) } ) ?? 0 {
                if let t = try? interp.output(at: i) {
                    if t.shape.dimensions.last == 16 { bfRegressorsIdx = i }
                    if t.shape.dimensions.last == 1  { bfScoresIdx = i }
                }
            }
        }
        return interp
    }

    // MARK: - Face Recognition

    @objc(runInference:candidatesJson:threshold:resolver:rejecter:)
    func runInference(
        _ faceFrameBase64: String,
        candidatesJson: String,
        threshold: Double,
        resolve: @escaping RCTPromiseResolveBlock,
        reject:  @escaping RCTPromiseRejectBlock
    ) {
        let t0 = Date()
        guard
            let imgData = Data(base64Encoded: faceFrameBase64),
            let uiImg = UIImage(data: imgData),
            let bmp = uiImg.resized(to: CGSize(width: mfnSize, height: mfnSize)),
            let interp = mfnInterpreter
        else {
            reject("INFERENCE_ERROR", "Could not decode image", nil); return
        }

        do {
            // Build INT8 input buffer
            let inputTensor = try interp.input(at: 0)
            let scale    = inputTensor.quantizationParameters?.scale    ?? 0.00784
            let zeroPoint = inputTensor.quantizationParameters?.zeroPoint ?? -1

            var inputBytes = [Int8](repeating: 0, count: mfnSize * mfnSize * 3)
            guard let pixelData = bmp.rgbPixelData() else {
                reject("INFERENCE_ERROR", "Could not get pixel data", nil); return
            }
            for i in 0 ..< mfnSize * mfnSize {
                let r = Float(pixelData[i * 3 + 0]) / 127.5 - 1.0
                let g = Float(pixelData[i * 3 + 1]) / 127.5 - 1.0
                let b = Float(pixelData[i * 3 + 2]) / 127.5 - 1.0
                inputBytes[i * 3 + 0] = quantize(r, scale: Float(scale), zp: zeroPoint)
                inputBytes[i * 3 + 1] = quantize(g, scale: Float(scale), zp: zeroPoint)
                inputBytes[i * 3 + 2] = quantize(b, scale: Float(scale), zp: zeroPoint)
            }
            try interp.copy(Data(bytes: inputBytes, count: inputBytes.count), toInputAt: 0)
            try interp.invoke()

            let outputTensor = try interp.output(at: 0)
            let queryEmb = outputTensor.data.withUnsafeBytes {
                Array(UnsafeBufferPointer<Float>(start: $0.baseAddress!.assumingMemoryBound(to: Float.self),
                                                count: embedDim))
            }
            let queryNorm = l2norm(queryEmb)

            // Compare candidates
            guard let jsonData = candidatesJson.data(using: .utf8),
                  let candidates = try? JSONSerialization.jsonObject(with: jsonData) as? [[String: Any]]
            else {
                reject("INFERENCE_ERROR", "Invalid candidates JSON", nil); return
            }

            var bestId: String? = nil
            var bestScore: Float = 0
            for entry in candidates {
                if entry["isRevoked"] as? Bool == true { continue }
                guard
                    let embB64 = entry["embeddingBase64"] as? String,
                    let embData = Data(base64Encoded: embB64)
                else { continue }
                let embFloat = embData.withUnsafeBytes {
                    Array(UnsafeBufferPointer<Float>(start: $0.baseAddress!.assumingMemoryBound(to: Float.self),
                                                    count: embData.count / 4))
                }
                let score = cosine(queryEmb, normA: queryNorm, b: embFloat)
                if score > bestScore {
                    bestScore = score
                    bestId = entry["workerId"] as? String
                }
            }

            let inferenceMs = Date().timeIntervalSince(t0) * 1000
            resolve([
                "workerId":    bestScore >= Float(threshold) ? bestId as Any : NSNull(),
                "confidence":  Double(bestScore),
                "inferenceMs": inferenceMs,
                "qualityScore": 1.0,
            ])
        } catch {
            reject("INFERENCE_ERROR", error.localizedDescription, error)
        }
    }

    // MARK: - Liveness Detection

    @objc(checkLiveness:challenge:resolver:rejecter:)
    func checkLiveness(
        _ framesBase64: [String],
        challenge: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject:  @escaping RCTPromiseRejectBlock
    ) {
        let t0 = Date()
        var earSeries  = [Float]()
        var yawSeries  = [Float]()

        for b64 in framesBase64 {
            guard
                let data = Data(base64Encoded: b64),
                let img  = UIImage(data: data),
                let kps  = runBlazeFace(img)
            else { continue }
            earSeries.append(kps.earProxy)
            yawSeries.append(kps.yaw)
        }

        let durationMs = Date().timeIntervalSince(t0) * 1000
        var result: [String: Any] = ["durationMs": durationMs]

        switch challenge {
        case "blink":
            let passed = detectBlink(earSeries)
            result["passed"] = passed
            result["ear"]    = earSeries.average
        case "turn_left":
            let passed = earSeries.isEmpty ? false :
                Float(yawSeries.filter { $0 < -yawTurnThresh }.count) / Float(yawSeries.count) >= yawFrameRatio
            result["passed"]     = passed
            result["yawDegrees"] = yawSeries.average
        case "turn_right":
            let passed = earSeries.isEmpty ? false :
                Float(yawSeries.filter { $0 > yawTurnThresh }.count) / Float(yawSeries.count) >= yawFrameRatio
            result["passed"]     = passed
            result["yawDegrees"] = yawSeries.average
        default:
            result["passed"] = false
        }
        resolve(result)
    }

    // MARK: - BlazeFace

    private struct FaceKps { let earProxy: Float; let yaw: Float }

    private func runBlazeFace(_ src: UIImage) -> FaceKps? {
        guard
            let bmp = src.resized(to: CGSize(width: bfSize, height: bfSize)),
            let pixels = bmp.rgbPixelData(),
            let interp = bfInterpreter
        else { return nil }

        var inputBytes = [Float](repeating: 0, count: bfSize * bfSize * 3)
        for i in 0 ..< bfSize * bfSize {
            inputBytes[i * 3 + 0] = Float(pixels[i * 3 + 0]) / 255.0
            inputBytes[i * 3 + 1] = Float(pixels[i * 3 + 1]) / 255.0
            inputBytes[i * 3 + 2] = Float(pixels[i * 3 + 2]) / 255.0
        }

        guard (try? interp.copy(Data(bytes: inputBytes, count: inputBytes.count * 4), toInputAt: 0)) != nil,
              (try? interp.invoke()) != nil,
              let regTensor  = try? interp.output(at: bfRegressorsIdx),
              let scoreTensor = try? interp.output(at: bfScoresIdx)
        else { return nil }

        let regressors = regTensor.data.withUnsafeBytes {
            Array(UnsafeBufferPointer<Float>(start: $0.baseAddress!.assumingMemoryBound(to: Float.self),
                                            count: 896 * 16))
        }
        let rawScores = scoreTensor.data.withUnsafeBytes {
            Array(UnsafeBufferPointer<Float>(start: $0.baseAddress!.assumingMemoryBound(to: Float.self),
                                            count: 896))
        }

        var bestScore = bfScoreThresh
        var bestIdx   = -1
        for i in 0 ..< 896 {
            let s = sigmoid(rawScores[i])
            if s > bestScore { bestScore = s; bestIdx = i }
        }
        guard bestIdx >= 0 else { return nil }

        return decodeKps(Array(regressors[(bestIdx * 16) ..< (bestIdx * 16 + 16)]),
                         anchor: bfAnchors[bestIdx])
    }

    private func decodeKps(_ raw: [Float], anchor: [Float]) -> FaceKps {
        let scale: Float = 128.0
        let faceW = max(raw[3] / scale, 0.01)
        let faceH = max(raw[2] / scale, 0.01)
        var kpX = [Float](repeating: 0, count: 6)
        var kpY = [Float](repeating: 0, count: 6)
        for i in 0 ..< 6 {
            kpX[i] = raw[4 + i * 2 + 1] / scale + anchor[0]
            kpY[i] = raw[4 + i * 2]     / scale + anchor[1]
        }
        let eyeMidY = (kpY[0] + kpY[1]) / 2
        let earProxy = min(max((kpY[3] - eyeMidY) / faceH, 0), 1)
        let yaw = ((kpX[2] - (kpX[0] + kpX[1]) / 2) / faceW) * 90
        return FaceKps(earProxy: earProxy, yaw: yaw)
    }

    private func detectBlink(_ series: [Float]) -> Bool {
        guard series.count >= 2 else { return false }
        var run = 0; var maxRun = 0
        for v in series {
            if v < blinkEarThresh { run += 1; maxRun = max(maxRun, run) } else { run = 0 }
        }
        return maxRun >= 2
    }

    private func generateBlazeFaceAnchors() -> [[Float]] {
        let strides = [8, 16]
        let apc = [2, 6]
        var anchors = [[Float]]()
        anchors.reserveCapacity(896)
        for s in strides.indices {
            let cells = bfSize / strides[s]
            for row in 0 ..< cells {
                for col in 0 ..< cells {
                    let cx = (Float(col) + 0.5) / Float(cells)
                    let cy = (Float(row) + 0.5) / Float(cells)
                    for _ in 0 ..< apc[s] { anchors.append([cx, cy]) }
                }
            }
        }
        return anchors
    }

    // MARK: - Math helpers

    private func quantize(_ v: Float, scale: Float, zp: Int) -> Int8 {
        let q = Int((v / scale).rounded()) + zp
        return Int8(max(-128, min(127, q)))
    }
    private func l2norm(_ v: [Float]) -> Float {
        max(sqrt(v.reduce(0) { $0 + $1 * $1 }), 1e-10)
    }
    private func cosine(_ a: [Float], normA: Float, b: [Float]) -> Float {
        let normB = l2norm(b)
        var dot: Float = 0
        for i in 0 ..< min(a.count, b.count) { dot += a[i] * b[i] }
        return max(-1, min(1, dot / (normA * normB)))
    }
    private func sigmoid(_ x: Float) -> Float { 1 / (1 + exp(-x)) }
}

// MARK: - UIImage helpers

private extension UIImage {
    func resized(to size: CGSize) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(size, true, 1.0)
        defer { UIGraphicsEndImageContext() }
        draw(in: CGRect(origin: .zero, size: size))
        return UIGraphicsGetImageFromCurrentImageContext()
    }

    func rgbPixelData() -> [UInt8]? {
        guard let cgImg = cgImage else { return nil }
        let w = cgImg.width; let h = cgImg.height
        var data = [UInt8](repeating: 0, count: w * h * 3)
        let ctx = CGContext(data: &data, width: w, height: h, bitsPerComponent: 8,
                           bytesPerRow: w * 3, space: CGColorSpaceCreateDeviceRGB(),
                           bitmapInfo: CGImageAlphaInfo.none.rawValue)
        ctx?.draw(cgImg, in: CGRect(x: 0, y: 0, width: w, height: h))
        return data
    }
}

private extension Array where Element == Float {
    var average: Double { isEmpty ? 0.0 : Double(reduce(0, +)) / Double(count) }
}
